src/dev/win32/lv_win32.c

```diff
@@ -0,0 +1,1260 @@
+    if(!ModuleHandle) {
+        return FALSE;
+    }
+
+    typedef BOOL(WINAPI * FunctionType)(HTOUCHINPUT);
+
+    FunctionType pFunction = (FunctionType)(
+                                 GetProcAddress(ModuleHandle, "CloseTouchInputHandle"));
+    if(!pFunction) {
+        return FALSE;
+    }
+
+    return pFunction(hTouchInput);
+}
+
+static UINT lv_win32_get_dpi_for_window(
+    _In_ HWND WindowHandle)
+{
+    UINT Result = (UINT)(-1);
+
+    HMODULE ModuleHandle = LoadLibraryW(L"SHCore.dll");
+    if(ModuleHandle) {
+        typedef enum MONITOR_DPI_TYPE_PRIVATE {
+            MDT_EFFECTIVE_DPI = 0,
+            MDT_ANGULAR_DPI = 1,
+            MDT_RAW_DPI = 2,
+            MDT_DEFAULT = MDT_EFFECTIVE_DPI
+        } MONITOR_DPI_TYPE_PRIVATE;
+
+        typedef HRESULT(WINAPI * FunctionType)(
+            HMONITOR, MONITOR_DPI_TYPE_PRIVATE, UINT *, UINT *);
+
+        FunctionType pFunction = (FunctionType)(
+                                     GetProcAddress(ModuleHandle, "GetDpiForMonitor"));
+        if(pFunction) {
+            HMONITOR MonitorHandle = MonitorFromWindow(
+                                         WindowHandle,
+                                         MONITOR_DEFAULTTONEAREST);
+
+            UINT dpiX = 0;
+            UINT dpiY = 0;
+            if(SUCCEEDED(pFunction(
+                             MonitorHandle,
+                             MDT_EFFECTIVE_DPI,
+                             &dpiX,
+                             &dpiY))) {
+                Result = dpiX;
+            }
+        }
+
+        FreeLibrary(ModuleHandle);
+    }
+
+    if(Result == (UINT)(-1)) {
+        HDC hWindowDC = GetDC(WindowHandle);
+        if(hWindowDC) {
+            Result = GetDeviceCaps(hWindowDC, LOGPIXELSX);
+            ReleaseDC(WindowHandle, hWindowDC);
+        }
+    }
+
+    if(Result == (UINT)(-1)) {
+        Result = LV_WIN32_DEFAULT_SCREEN_DPI;
+    }
+
+    return Result;
+}
+
+static void lv_win32_display_driver_flush_callback(
+    lv_display_t * disp_drv,
+    const lv_area_t * area,
+    uint8_t * px_map)
+{
+    HWND window_handle = (HWND)lv_display_get_user_data(disp_drv);
+    if(!window_handle) {
+        lv_display_flush_ready(disp_drv);
+        return;
+    }
+
+    lv_win32_window_context_t * context = (lv_win32_window_context_t *)(
+                                              lv_win32_get_window_context(window_handle));
+    if(!context) {
+        lv_display_flush_ready(disp_drv);
+        return;
+    }
+
+    if(lv_display_flush_is_last(disp_drv) && !context->display_refreshing) {
+#if (LV_COLOR_DEPTH == 32) || \
+    (LV_COLOR_DEPTH == 24) || \
+    (LV_COLOR_DEPTH == 16 && LV_COLOR_16_SWAP == 0) || \
+    (LV_COLOR_DEPTH == 8)
+        UNREFERENCED_PARAMETER(px_map);
+        memcpy(
+            context->display_framebuffer_base,
+            context->display_draw_buffer_base,
+            context->display_draw_buffer_size);
+#elif (LV_COLOR_DEPTH == 16 && LV_COLOR_16_SWAP != 0)
```

Comment:
`LV_COLOR_16_SWAP` is removed from v9.

Reply:
I think I will remove that in the simulator.

Kenji Mouri

Reply:
Ok, then I'll refer to the SDL and make some modifications